### PR TITLE
feat: update stack

### PR DESCRIPTION
**Pull request by Eikix**
## Pull Request type

Please check the type of change your PR introduces:

- [ ] Bugfix
- [x] Feature
- [ ] Code style update (formatting, renaming)
- [ ] Refactoring (no functional changes, no API changes)
- [ ] Build-related changes
- [ ] Documentation content changes
- [ ] Other (please describe):

## What is the current behavior?

Resolves: #373 

## Does this introduce a breaking change?

- [x] Yes
- [ ] No

### CODE REVIEW

File: crates/evm/src/stack.cairo [hunk 1]
```diff
@@ -24,14 +24,18 @@ use evm::helpers::U256IntoResultU32;
 use starknet::EthAddress;
 use utils::i256::i256;
 
+const STACK_SEGMENT_SIZE: usize = 1024;
```
Comment (Eikix) [inline crates/evm/src/stack.cairo:27]: could also take directly constants::STACK_MAX_DEPTH, no strong opinion

File: crates/evm/src/stack.cairo [hunk 2]
```diff
@@ -74,12 +93,13 @@ impl StackImpl of StackTrait {
...
     fn pop(ref self: Stack) -> Result<u256, EVMError> {
-        if self.len() == 0 {
+        let length = self.len();
+        if length == 0 {
             return Result::Err(EVMError::StackError(STACK_UNDERFLOW));
         }
-        let last_index = self.len() - 1;
-        self.len -= 1;
-        let item = self.items.get(last_index.into());
+        self.len.insert(self.active_segment().into(), length - 1);
```
Comment (Eikix) [inline crates/evm/src/stack.cairo:100]: i could technically abstract a setter for the length, no strong opinion, this feels fine imo

File: crates/evm/src/stack.cairo [hunk 3]
```diff
@@ -198,13 +218,26 @@ impl StackImpl of StackTrait {
 
     /// Returns the length of the stack.
     #[inline(always)]
-    fn len(self: @Stack) -> usize {
-        *self.len
+    fn len(ref self: Stack) -> usize {
```
Comment (Eikix) [inline crates/evm/src/stack.cairo:221]: accessing a feltdict needs a mutable ref

File: crates/evm/src/stack.cairo [hunk 4]
```diff
@@ -51,19 +57,32 @@ impl StackImpl of StackTrait {
...
     fn push(ref self: Stack, item: u256) -> Result<(), EVMError> {
+        let length = self.len();
         // we can store at most 1024 256-bits words
-        if self.len() == constants::STACK_MAX_DEPTH {
+        if length == constants::STACK_MAX_DEPTH {
             return Result::Err(EVMError::StackError(STACK_OVERFLOW));
         }
-        self.items.insert(self.len.into(), NullableTrait::new(item));
-        self.len += 1;
+        let index = self.compute_active_segment_index(self.len());
```
Comment (enitrat) [inline crates/evm/src/stack.cairo:83]: use pre-cached `length` variable?

File: crates/evm/src/stack.cairo (hunk 2 above)
Comment (enitrat) [reply crates/evm/src/stack.cairo:100]: fine for me given that it's accessed internally

**Review by enitrat — COMMENTED** [review]
v good

**Review by enitrat — APPROVED** [review]

**Pull request merged**